io/pci/pci_hotplug.py

```diff
@@ -47,76 +47,82 @@ def setUp(self):
+                            slot).group()
         else:
-            self.slot = pci.get_pci_prop(self.device, "PhySlot")
-        if not os.path.isdir('/sys/bus/pci/slots/%s' % self.slot):
-            self.cancel("%s Slot not available" % self.slot)
-        if not os.path.exists('/sys/bus/pci/slots/%s/power' % self.slot):
-            self.cancel("%s Slot does not support hotplug" % self.slot)
+            slot = pci.get_pci_prop(pci_addr, "PhySlot")
+        if not os.path.isdir('/sys/bus/pci/slots/%s' % slot):
+            self.cancel("%s Slot not available" % slot)
+        if not os.path.exists('/sys/bus/pci/slots/%s/power' % slot):
+            self.cancel("%s Slot does not support hotplug" % slot)
+        return slot
 
     def test(self):
         """
         Creates namespace on the device.
         """
+        for pci_addr in self.device.split(","):
+            if not os.path.isdir('/sys/bus/pci/devices/%s' % pci_addr):
+                self.cancel(" %s does not exist" % pci_addr)
+            slot = self.get_slot(pci_addr)
+            for _ in range(self.count):
+                self.hotplug_remove(slot, pci_addr)
+                self.hotplug_add(slot, pci_addr)
+                self.check_add_remove(pci_addr)
 
-        for _ in range(self.num_of_hotplug):
-            self.hotplug_remove()
-            self.hotplug_add()
-            self.check_add_remove()
-
-    def hotplug_remove(self):
+    def hotplug_remove(self, slot, pci_addr):
         """
         Hot Plug remove operation
         """
-        genio.write_file("/sys/bus/pci/slots/%s/power" % self.slot, "0")
+        genio.write_file("/sys/bus/pci/slots/%s/power" % slot, "0")
         time.sleep(5)
-        if self.device in pci.get_pci_addresses():
+        if pci_addr in pci.get_pci_addresses():
             self.return_code = 1
         else:
-            self.log.info("Adapter %s removed successfully", self.device)
+            self.log.info("Adapter %s removed successfully", pci_addr)
 
-    def hotplug_add(self):
+    def hotplug_add(self, slot, pci_addr):
         """
         Hot plug add operation
         """
-        genio.write_file("/sys/bus/pci/slots/%s/power" % self.slot, "1")
+        genio.write_file("/sys/bus/pci/slots/%s/power" % slot, "1")
         time.sleep(5)
-        if self.device not in pci.get_pci_addresses():
+        if pci_addr not in pci.get_pci_addresses():
             self.return_code = 2
         else:
-            self.log.info("Adapter %s added back successfully", self.device)
+            self.log.info("Adapter %s added back successfully", pci_addr)
 
-    def check_add_remove(self):
+    def check_add_remove(self, pci_addr):
         """
         Function to check if the adapter is removed and
         added back as desired
         """
         if self.return_code == 1:
-            self.fail('%s not removed' % self.device)
+            self.fail('%s not removed' % pci_addr)
```

Comment:
@Naresh-ibm Since you are handling the test for multiple pci addresses, you need to store the failed ones and fail at the last.

Reply:
sure, this I will change.

Reply:
@Naresh-ibm Is this not taken care of yet ?